add/bugfix: new paintkit for welding

endipe opened the pull request:
## Описание
add:
- Добавляет в тир 2+ лодаут 2 пейнткита для сварочных масок, "Big brother" и "Slavic", отличаются данные "скины" от раскрасок что уже есть в игре тем, что данные - одноразовые, то есть ли вы или кто-либо еще помоет вашу дриповую маску - то все, остается только плакать.

bugfix:
- Исправляет ссылку у греев на иконки. Без этого кастомные версии сварочной маски не отображаются у греев.
- Исправляет возможность сделать бесконечный баллончик краски, если покрасить сварочную маску при единице заряда.
- Теперь при покраске иконка на персонаже обновляется сразу.

## Ссылка на предложение/Причина создания ПР
https://discord.com/channels/617003227182792704/755125334097133628/1249705481564327978 — прошло голосование
Автор предложения: **splitpussy**

## Демонстрация изменений

![dreamseeker_QpVycTYRVc](https://github.com/user-attachments/assets/ae84b4e0-7503-4c72-bd5d-9a75ebbe00f5)
![kh5I0XmCtZ](https://github.com/user-attachments/assets/92f22247-335d-4f60-a0b6-7cc354cdcdd6)

endipe committed:
add paintkit for welding

+ bugfix

endipe committed:
Update loadout_donor.dm

fix path bigbrother and display_name

endipe committed:
polishing

убрано лишнее + маленькие изменения

endipe committed:
fix name

bigbrother!!!

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -328,3 +327,27 @@
+
+/obj/item/toy/crayon/spraycan/paintkit/New()
+	..()
+	name = "Paintkit '[name]'"
```
нет, для апдейта имени мы используем update_name()

и используй Initialize, вместо New. в оригинале также поменяйю

Gottfrei replied:
хотя у тебя можно не использовать update_name(), потому что меняется только на инициализации.

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -68,25 +78,35 @@
-		var/obj/item/toy/crayon/spraycan/C = I
+		var/obj/item/toy/crayon/spraycan/paintkit/C = I
 		if(C.capped)
-			to_chat(user, "<span class = 'warning'>Вы не можете раскрасить [src], если крышка на банке закрыта!</span>")
+			to_chat(user, span_warning("Вы не можете раскрасить [src], если крышка баллона краски закрыта!"))
+			return
+		if(C.uses <= 1)
+			to_chat(user, span_warning("Не похоже, что бы осталось достаточно краски"))
 			return
 		var/list/weld_icons = list("Flame" = image(icon = src.icon, icon_state = "welding_redflame"),
 									"Blue Flame" = image(icon = src.icon, icon_state = "welding_blueflame"),
 									"White Flame" = image(icon = src.icon, icon_state = "welding_white"))
 		var/list/weld = list("Flame" = "welding_redflame",
 							"Blue Flame" = "welding_blueflame",
 							"White Flame" = "welding_white")
-		var/choice = show_radial_menu(user, src, weld_icons)
+		var/choice = null
+		if(istype(I, /obj/item/toy/crayon/spraycan/paintkit))
+			if(C.used)
+				to_chat(user, span_warning("Кажется, этот набор уже был использован. Вам этого не хватит для покраски."))
+				return
+			weld = list(C.name = C.weld_icons)
+			choice = C.name
+			C.used = TRUE
+		if(!choice)
+			choice = show_radial_menu(user, src, weld_icons)
```
здесь происходят какие-то максимально странные нон-ООП манипуляции.

endipe replied:
окей, перепишу

endipe replied:
лучше?

endipe committed:
finally?

replace New() -> Initialize
update attackby()

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -328,3 +327,27 @@
+	var/weld_icons = "welding" //default
+
+/obj/item/toy/crayon/spraycan/paintkit/Initialize(mapload)
+	..()
```
```suggestion
	. = ..()
```

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -328,3 +327,27 @@
+
+/obj/item/toy/crayon/spraycan/paintkit/Initialize(mapload)
+	..()
+	name = "Paintkit «[name]»"
```
зачем мы вообще меняем имя на инициализации, если оно всегда одинаковое?

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,17 +60,31 @@
+	if(istype(I, /obj/item/toy/crayon/spraycan/paintkit))
+		if(!check_spraycan(I, user))
 			return
-		var/obj/item/toy/crayon/spraycan/C = I
-		if(C.capped)
-			to_chat(user, "<span class = 'warning'>Вы не можете раскрасить [src], если крышка на банке закрыта!</span>")
+		var/obj/item/toy/crayon/spraycan/paintkit/C = I
+		if(C.used)
+			to_chat(user, span_warning("Кажется, этот набор уже был использован. Вам этого не хватит для покраски."))
+			return
+		C.used = TRUE
+		adjust_paint(C.weld_icons, user, I)
+	else if(istype(I, /obj/item/toy/crayon/spraycan))
+		if(!check_spraycan(I, user))
```
весь это блок можно перенести в отдельный прок покраски, если по маске тыкнули баллончиком.

Gottfrei replied:
также я не вижу причин почему не сделать всем баллончикам ориентацию на количество использований. но у базовых это число будет бесконечным.

endipe replied:
> также я не вижу причин почему не сделать всем баллончикам ориентацию на количество использований. но у базовых это число будет бесконечным.

В целом тоже так подумал, но не хотел рефакторить старый код... но в итоге реально минорный рефактор вышел

endipe committed:
It's over?

- paintkit/Initialize(mapload) removed.
- reworked attackby for obj/welding
- spraycan refactoring

endipe commented:
думаю, что так лучше получилось... ну или я создал новые проблемы, idk

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -269,16 +269,20 @@
 	icon_state = "spraycan_cap"
 	desc = "A metallic container containing tasty paint."
 	var/capped = 1
+	var/used = FALSE
```
просто используй -1 для бесконечного, вместо двух переменных.

endipe replied:
> просто используй -1 для бесконечного, вместо двух переменных.

окей, сделаю уже завтра

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,41 +60,50 @@
+	else return ..()
+
+/obj/item/clothing/head/welding/proc/adjust_paint(var/weld_icon_state, mob/living/user = null, var/obj/item/toy/crayon/spraycan/spray = null)
+	icon_state = weld_icon_state
+	if(weld_icon_state == initial(icon_state))
 		paint = null
```
это плохо.

endipe replied:
> это плохо.

плохо что initial(icon_state)? Или прямое назначение icon_state от weld_icon_state?

endipe replied:
переписал, теперь идет только paint

endipe committed:
before they run out...

endipe committed:
oops

fix uses, rename update_not_used and is_can_draw

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,41 +61,50 @@
+	icon_state = paint ? paint : base_icon_state
+	return ..()
 
+/obj/item/clothing/head/welding/proc/adjust_paint(var/weld_icon_state, mob/living/user = null, var/obj/item/toy/crayon/spraycan/spray = null)
```
var не нужны в аргументах

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -328,3 +331,38 @@
+	not_used = (not_used > 0) ? FALSE : not_used
+	uses--
+
+/obj/item/toy/crayon/spraycan/proc/is_can_draw_paint(var/obj/object, mob/living/user)
```
```suggestion
/obj/item/toy/crayon/spraycan/proc/can_paint(var/obj/object, mob/living/user)
```
плохое название.

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -269,16 +269,19 @@
 	icon_state = "spraycan_cap"
 	desc = "A metallic container containing tasty paint."
 	var/capped = 1
+	var/not_used = -1 // -1 for unlimited used
```
```suggestion
	var/uses = -1 // -1 for unlimited used
```
плохое название. также всем переменным нужен автодок.

Gottfrei commented on code/game/objects/items/crayons.dm:
```diff
@@ -328,3 +331,38 @@
+		return FALSE
+	return TRUE
+
+/obj/item/toy/crayon/spraycan/paintkit
```
прок attack_self не оверрайднут.

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,41 +61,50 @@
-		var/obj/item/toy/crayon/spraycan/C = I
-		if(C.capped)
-			to_chat(user, "<span class = 'warning'>Вы не можете раскрасить [src], если крышка на банке закрыта!</span>")
+		if(!spray.is_can_draw_paint(src, user))
```
это всё нужно перенести в прок покраски. здесь должна быть только проверка на тип и вызов прока покраски.

endipe committed:
end?

endipe committed:
Merge branch 'master220' into add-paintkit-for-weld

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,41 +62,49 @@
+			weld_icons += list("[weld_icon]" = image(icon = src.icon, icon_state = spray.weld_icons[weld_icon]))
 		var/choice = show_radial_menu(user, src, weld_icons)
-		if(!choice || I.loc != user || !Adjacent(user))
+		if(!choice || spray.loc != user || !Adjacent(user))
```
если ты уже проверяешь лок, то проверка на близость не требуется.

endipe replied:
> если ты уже проверяешь лок, то проверка на близость не требуется.

не трогал проверку, она осталась от прошлого кодера

Gottfrei commented on code/modules/clothing/head/misc_special.dm:
```diff
@@ -60,41 +62,49 @@
+			weld_icons += list("[weld_icon]" = image(icon = src.icon, icon_state = spray.weld_icons[weld_icon]))
 		var/choice = show_radial_menu(user, src, weld_icons)
-		if(!choice || I.loc != user || !Adjacent(user))
+		if(!choice || spray.loc != user || !Adjacent(user))
```
```suggestion
		if(!choice || spray.loc != user)
```

Gottfrei committed:
No Adjacency Check

Gottfrei commented:
в игре потестил всё?

endipe commented:
да, всё работало как ожидалось

endipe commented:
единственное, что смущает - это изменение в уменьшение размера файла некоторых спрайтов, но icon dif ничего не говорит

![изображение](https://github.com/user-attachments/assets/21004e51-79d9-48cf-854e-65d0b236cb7f)
